### DISCUSSION TITLE
[Subset] about pyftsubset's --no-canonical-order option

### DISCUSSION THREAD

**Discussion by yisibl** [harfbuzz/harfbuzz#4178] [Q&A] [answered]
> --no-canonical-order
Keep original order of font tables. This is faster. [default]
https://fonttools.readthedocs.io/en/latest/subset/index.html

Is the default behavior of `hb-subset` consistent with this option?

Also, what other options are available to speed things up without considering the size of the generated subset file?

**Comment by behdad** [comment]
> > [](https://fonttools.readthedocs.io/en/latest/subset/index.html)
> 
> Is the default behavior of `hb-subset` consistent with this option?

I think `hb-subset` doesn't try to sort tables canonically. At any rate, it wouldn't have any significant speed difference.

> Also, what other options are available to speed things up without considering the size of the generated subset file?

If you are subsetting the same face repeatedly you can use `hb_subset_preprocess()` which would significantly speed up subsetting.

**Reply by garretrieger** [reply, marked as answer]
hb-subset by default will sort tables from smallest to largest size in bytes. It's possible to override the ordering using the API (https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-face.h#L181) but not via the command line util. As Behdad mentioned the sort order should have no measurable impact on performance.

Preprocessing is the main way to speed up subsetting, but that will only yield benefits when amortized over multiple subsetting operations.

--no-prune-unicode-range may yield small performance improvements, but I suspect not much.

If the font in question has a large GSUB/GPOS table you may be able to gain speedups by disabling layout subsetting. This can be done by adding GSUB/GPOS/GDEF to the no subset table list and set the retain gids option. This will leave the layout tables untouched and since retain gids is set, the gids in the tables will still be correct.
